realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java

```diff
@@ -3705,5 +3867,80 @@ public void limit_invalidValuesThrows() {
         }
     }
 
+    // FIXME Maybe move to QueryDescriptor or maybe even to RealmFieldType?
+    private boolean supportDistinct(RealmFieldType type) {
+        switch (type) {
+            case INTEGER:
+            case BOOLEAN:
+            case STRING:
+            case BINARY:
+            case DATE:
+            case FLOAT:
+            case DOUBLE:
+            case OBJECT:
+            case DECIMAL128:
+            case OBJECT_ID:
+            case LINKING_OBJECTS:
+                return true;
+            case LIST:
+            case INTEGER_LIST:
+            case BOOLEAN_LIST:
+            case STRING_LIST:
+            case BINARY_LIST:
+            case DATE_LIST:
+            case FLOAT_LIST:
+            case DOUBLE_LIST:
+            case DECIMAL128_LIST:
+            case OBJECT_ID_LIST:
+                return false;
+        }
+        // Should never reach here as the above switch is exhaustive
+        throw new UnsupportedOperationException("Unhandled realm field type " + type);
+    }
+
+    // Manual distinct method for verification. Uses field value's equals.
+    @NotNull
+    private Set<List<? super Object>> distinct(RealmResults<DynamicRealmObject> all, Object... fields) {
+        Set<List<? super Object>> values = new HashSet();
+
+        // Parsed hierarchical field accessors
+        List<String[]> fieldAccessors = new ArrayList<>();
+        for (Object field : fields) {
+            fieldAccessors.add(((String) field).split("\\."));
+        }
+
+        for (DynamicRealmObject object : all) {
+            List<? super Object> elements = new ArrayList<>(fields.length);
+            for (String[] split : fieldAccessors) {
+                int i = 0;
+                while(i < split.length - 1) {
+                    object = object.get(split[i]);
+                    i++;
+                }
+                String fieldName = split[i];
+                if (!object.isNull(fieldName)) {
+                    Object e = object.get(fieldName);
+                    // Need to convert byte arrays to list to detect duplicates when inserting to values
+                    if (e instanceof byte[]) {
+                        elements.add(convertBytesToList((byte[]) e));
+                    } else {
+                        elements.add(e);
+                    }
+                } else {
+                    elements.add(NULL_PLACEHOLDER);
```

Comment:
Not sure I understand this?

Reply:
Keeping track of null values as they are still considered distinct. Previous implementation used container not allow null values directly. The current list allows null values, so eliminated the `NULL_PLACEHOLDER`.